### ISSUE TITLE
strange / wrong bucket name in s3 policy in authRole

### ISSUE THREAD

**Issue by AndreasEK** [aws-amplify/amplify-cli#9858]
### Before opening, please confirm:

- [X] I have installed the latest version of the Amplify CLI (see above), and confirmed that the issue still persists.
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-cli/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-cli/blob/master/CONTRIBUTING.md#bug-reports).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

14.16.1

### Amplify CLI Version

7.6.21

### What operating system are you using?

Mac

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

not that I'm aware of O:-)

### Amplify Categories

storage

### Amplify Commands

Not applicable

### Describe the bug

I am migrating from amplify-cli 6.4.* to the latest, and I just realised that the s3 related policies in authRole contain the wrong bucket name. The amplify environment is rendered in twice, in a particular weird way:

```
{
    "Version": "2012-10-17",
    "Statement": {
        "Action": [
            "s3:PutObject",
            "s3:GetObject",
            "s3:DeleteObject"
        ],
        "Resource": [
            "arn:aws:s3:::solinger-sportbund-images-bucketdev-dev/protected/${cognito-identity.amazonaws.com:sub}/*"
        ],
        "Effect": "Allow"
    }
}
```

cli-inputs.json: (specifies correct bucketName)
```
{
  "resourceName": "AmplifySolingerSportbundImagesRN",
  "bucketName": "solinger-sportbund-images-bucket",
  "policyUUID": "f330379c",
  "storageAccess": "authAndGuest",
  "guestAccess": [
    "READ"
  ],
  "authAccess": [
    "CREATE_AND_UPDATE",
    "READ",
    "DELETE"
  ]
}
```

local-env-info.json: (constains envName)
```
{
  "projectPath": "/Users/aek/dev/ssbsg/solinger-sportbund-api",
  "defaultEditor": "vscode",
  "envName": "dev"
}
```

amplify-meta.json: (contains wrong bucket name in output)
```
  "storage": {
    "AmplifySolingerSportbundImagesRN": {
      "service": "S3",
      "providerPlugin": "awscloudformation",
      "dependsOn": [],
      "output": {
        "BucketName": "solinger-sportbund-images-bucketdev-dev",
        "Region": "eu-central-1"
      },
      "lastPushTimeStamp": "2022-02-25T12:16:48.966Z",
      "providerMetadata": {
        "s3TemplateURL": "https://s3.amazonaws.com/amplify-solinger-v1-dev-20190523135333-deployment/amplify-cfn-templates/storage/cloudformation-template.json",
        "logicalId": "storageAmplifySolingerSportbundImagesRN"
      },
      "lastPushDirHash": "eGT60r2X8AS6C+FfYZ2Gk9JBHZE="
    }
  },
```

### Expected behavior

The resource should be:

```
arn:aws:s3:::solinger-sportbund-images-bucket-dev/protected/${cognito-identity.amazonaws.com:sub}/*
```

### Reproduction steps

1. not sure - I assume a new project doesn't show this behaviour, but reproducing it with a legacy project might be difficult?

### GraphQL schema(s)

<details>

```graphql
# Put schemas below this line


```

</details>

### Log output

<details>

```
# Put your logs below this line


```

</details>

**Comment by AndreasEK** [comment]
So this appears to be the part that is producing the wrong value:

```
        "BucketName": {
          "Fn::If": [
            "ShouldNotCreateEnvResources",
            {
              "Ref": "bucketName"
            },
            {
              "Fn::Join": [
                "",
                [
                  {
                    "Ref": "bucketName"
                  },
                  {
                    "Fn::Select": [
                      3,
                      {
                        "Fn::Split": [
                          "-",
                          {
                            "Ref": "AWS::StackName"
                          }
                        ]
                      }
                    ]
                  },
                  "-",
                  {
                    "Ref": "env"
                  }
                ]
              ]
            }
          ]
        },
       ```

It joins these parts to the BucketName:

1. The `bucketName`
2. The 4th element of `AWS::StackName`
3. "-"
4. The environmant name

1. resolves to solinger-sportbund-images-bucket
2. the root stack is called amplify-solinger-v1-dev-20190523135333, if this is split by '-' the 4th element (3 with zero index) is indeed 'dev'
3. '-'
4. the environement is 'dev'

So this resolves in solinger-sportbund-images-bucketdev-dev - but that's not what the bucket is called. Well … it actually has been created unintentionally!
<img width="1288" alt="image" src="https://user-images.githubusercontent.com/1392994/155732603-0839f9a3-d6a4-41af-bcfe-bbbc5e738e0d.png">

The question is now - how do I fix that?

**Comment by AndreasEK** [comment]
So, it looks like the naming pattern was introduced in https://github.com/aws-amplify/amplify-cli/pull/2709/files#diff-3b211d52016482ec677f83b3c045e3791847b9f9fec3c2e25974e5a712c43c7f 

Our amplify project is older. 

Any idea how to fix our setup? Since the product is in production, re-creating the s3 bucket is not an option. otoh, I don't know how to force the BucketName to the old naming pattern.

**Comment by AndreasEK** [comment]
Sorry for all the comments, but while waiting on an answer I still tried to understand the situation better. 

I tried this in `override.ts`:
```
import { AmplifyS3ResourceTemplate } from '@aws-amplify/cli-extensibility-helper';

// AmplifyS3ResourceTemplate | see https://github.com/aws-amplify/amplify-cli/issues/9063#issuecomment-1034426494
export function override(resources: any) {
    resources.s3Bucket.bucketName = {
        "Fn::If": [
          "ShouldNotCreateEnvResources",
          {
            "Ref": "bucketName"
          },
          {
            "Fn::Join": [
              "",
              [
                {
                  "Ref": "bucketName"
                },
                "-",
                {
                  "Ref": "env"
                }
              ]
            ]
          }
        ]
      }
    resources.s3AuthProtectedPolicy.policyDocument.Statement.Action = 
        ["s3:PutObject","s3:GetObject","s3:DeleteObject","s3:PutObjectAcl"]
}
```

but when trying to push this, CloudFormation complains that the bucket already exists. Where is the decision made to recreate the resource?

```
UPDATE_IN_PROGRESS          S3Bucket                                                                                     AWS::S3::Bucket            Fri Feb 25 2022 17:19:33 GMT+0100 (Mitteleuropäische Normalzeit) Requested update requires the creation of a new physical resource; hence creating one.
UPDATE_FAILED               S3Bucket                                                                                     AWS::S3::Bucket            Fri Feb 25 2022 17:19:33 GMT+0100 (Mitteleuropäische Normalzeit) solinger-sportbund-images-bucket-dev already exists                                   
UPDATE_ROLLBACK_IN_PROGRESS amplify-solinger-v1-dev-20190523135333-storageAmplifySolingerSportbundImagesRN-1LU637KOZH4RX AWS::CloudFormation::Stack Fri Feb 25 2022 17:19:34 GMT+0100 (Mitteleuropäische Normalzeit) The following resource(s) failed to update: [S3Bucket].                               
⠧ Updating resources in the cloud. This may take a few minutes...

UPDATE_FAILED storageAmplifySolingerSportbundImagesRN AWS::CloudFormation::Stack Fri Feb 25 2022 17:19:35 GMT+0100 (Mitteleuropäische Normalzeit) Embedded stack arn:aws:cloudformation:eu-central-1:498438122852:stack/amplify-solinger-v1-dev-20190523135333-storageAmplifySolingerSportbundImagesRN-1LU637KOZH4RX/dcf19180-7d52-11e9-8dab-06537af0a09a was not successfully updated. Currently in UPDATE_ROLLBACK_IN_PROGRESS with reason: The following resource(s) failed to update: [S3Bucket]. 

```

It seems like there have been others with this issue (sorry again, that I didn't find the issue earlier): https://github.com/aws-amplify/amplify-cli/issues/8893#issuecomment-973029867

The resolution there however is to stay with the "new" bucket name and move/copy the contents over. 

@ray415 & @kaustavghosh06 When I search for "AWS::StackName" in my project I find that's all over the place, and I fear that this issue might hit me in other places, that I have not realized yet. Do you have any recommendations for me? Thanks a lot!

**Comment by ray415** [comment]
I copied the files from the bucket to another one and then deleted the initial bucket. After a push, the bucket was recreated and I copied the files back. You can probably test it with another test-environment first.

> Do you have any recommendations for me?

Migrate away from Amplify. I do not even dare to upgrade the Amplify version anymore because every time I do so it breaks something and this is not acceptable for the production environments to wait the fix for two weeks .

**Comment by AndreasEK** [comment]
I managed to get the situation resolved by removing and re-attaching the bucket via cloudformation resource imports. However, it was a bit nerve wrecking, I have to admin. 

@ray415 I am indeed considering the same. Today, I learned about `amplify status -v` which tells you, which resources *will be deleted*. As fragile as amplify is, this will at least help to gauge the risk before doing the actual push.

**Comment by ykethan** [comment]
Hello @AndreasEK, Tired to replicate the issue in amplify application by migrating from 6.4.0 to the latest but did not run into the issue. 

Following up to see if you are facing any issues in the latest CLI version.

**Comment by AndreasEK** [comment]
Hi @ykethan - as commented in https://github.com/aws-amplify/amplify-cli/issues/9858#issuecomment-1057126673 I was able to resolve the situation for my setup. After that I was just happy that it worked for me and have not investigated further.

**Comment by ykethan** [comment]
Hey @AndreasEK, glad to hear that you are up and running. Closing the issue, please do reach out to us again.

**Comment by irothenbaum** [comment]
> but when trying to push this, CloudFormation complains that the bucket already exists. Where is the decision made to recreate the resource?

Faced the same issue. After finally realizing the cause was the bucket name changed and implementing your override script to revert it, I encountered the same Already Exists error. Rather than trying to tell Amplify to skip creation, I tried just deleting the old bucket using the AWS Console and that worked! (see steps below if useful) So while perhaps a little scary, it actually only took a couple minutes to resolve. No update to client interfaces or integrations required.

Exact Steps:
1. Implement override script
    a. `amplify override storage` -> paste @AndreasEK 's script into `override.ts`
2. Go to AWS Console and Create a temporary bucket
3. Copy all objects from old bucket to temp bucket
  a. Using AWS CLI, run command: `aws s3 sync s3://<OLD BUCKET NAME> s3://<TEMPORARY BUCKET NAME>`
4. In AWS Console, delete the old bucket
  a. Will need to remove all objects first, but console walks you through this pretty well.
5. Then run `amplify push` - should no longer have the Already Exists error
6. Copy all files back to the new [old] bucket using  `aws s3 sync  s3://<TEMPORARY BUCKET NAME> s3://<OLD BUCKET NAME>`
